dbms/src/AggregateFunctions/AggregateFunctionIncrementalClustering.h

```diff
@@ -0,0 +1,560 @@
+            {
+                add_from_vector(rhs.clusters[i].center());
+            }
+            return;
+        }
+        if (clusters.size() < multiplication_factor * clusters_num)
+        {
+            std::vector<Cluster> old_clusters = std::move(clusters);
+            clusters = rhs.clusters;
+            for (size_t i = 0; i != old_clusters.size(); ++i)
+            {
+                add_from_vector(old_clusters[i].center());
+            }
+            return;
+        }
+
+        /// merge ближайших пар кластеров жадным алгоритмом
+        for (size_t i = 0; i != rhs.clusters.size(); ++i)
+        {
+            size_t closest_cluster = 0;
+            Float64 min_dist = compute_distance(clusters[i].center(), rhs.clusters[i].center());
+            for (size_t j = i + 1; j != clusters.size(); ++j)
+            {
+                Float64 cur_dist = compute_distance(clusters[j].center(), rhs.clusters[i].center());
+                if (cur_dist < min_dist)
+                {
+                    min_dist = cur_dist;
+                    closest_cluster = j;
+                }
+            }
+
+            clusters[closest_cluster].merge_cluster(rhs.clusters[i]);
+            std::swap(clusters[closest_cluster], clusters[i]);
+        }
+    }
+
+    void write(WriteBuffer & buf) const
+    {
+        writeBinary(clusters_num, buf);
+        writeBinary(dimensions, buf);
+        // writeBinary(initialized_clusters, buf);
+        for (auto & cluster : clusters)
+        {
+            cluster.write(buf);
+        }
+    }
+
+    void read(ReadBuffer & buf)
+    {
+        readBinary(clusters_num, buf);
+        readBinary(dimensions, buf);
+        // readBinary(initialized_clusters, buf);
+        for (auto & cluster : clusters)
+        {
+            cluster.read(buf);
+        }
+    }
+
+    Float64 predict(const std::vector<Float64> & predict_feature) const
+    {
+        /// Можно запилить обычную кластеризацию на малом числе элементов
+
+        size_t closest_cluster = 0;
+        Float64 min_dist = compute_distance(clusters[0].center(), predict_feature);
+        for (size_t i = 1; i != clusters.size(); ++i)
+        {
+            Float64 cur_dist = compute_distance(clusters[i].center(), predict_feature);
+            if (cur_dist < min_dist)
+            {
+                min_dist = cur_dist;
+                closest_cluster = i;
+            }
+        }
+
+        std::cout << "\n\nmy prediction: " << closest_cluster << "\n";
+        std::cout << "with min dist: " << min_dist << "\n";
+        for (size_t j = 0; j != clusters[closest_cluster].center().size(); ++j)
+        {
+            std::cout << clusters[closest_cluster].center()[j] << " ";
+        }
+        std::cout << "\n\n";
+
+        // TODO
+        std::ignore = predict_feature;
+        return 0;
+    }
+
+    void print_clusters(ColumnVector<Float64>& column) const
+    {
+        std::cout << "\n\nIM GOING TO PRINT CLUSTERS\n\n";
```

Comment:
Logger everywhere, instead of direct stream manipulation.

Reply:
It is a draft. This is not going to exist in final version.